### DISCUSSION TITLE
Meta Tags does not appear in View Source Code - Facebook Crawler issue - SEO

### DISCUSSION THREAD

**Discussion by desaijay315** [vercel/next.js#14277] [Help] [answered]
I have integrated the meta tags which are fetched dynamically via an API call. 

**Here is my component structure -** 

**pages/index.js**

`import {seoAPI} from '../action/api-handler';
`
`import Head from 'next/head';
````
class Index extends Component {

    constructor(props){
        super(props);
        const {seoData} = props;

        this.state = {
            seoData: seoData
        }
    }

    componentDidMount(){
        
      }

    render(){
        return (
            <>
               <Head>
<link rel="icon" href="/favicon.ico" />
</Head>
                <BaseLayout seo_arr={this.state.seoData}>
                </BaseLayout>
            </>
        )
    }
}
```



```
Index.getInitialProps = async () => {

    const seoData = await seoAPI();
  
    return {
        
        seoData
    }
}


export default Index;

```


**components/BaseLayout.js**

```
import Head from 'next/head';
import Header from './Header';
import Footer from './Footer';



```
```
const BaseLayout = (props) => {
    const { children, seo_arr } = props;

    const seoData = seo_arr


```    return(
        <>
        <Head>
            <meta httpEquiv="Content-Type" content="text/html; charset=utf-8" />
            <meta httpEquiv="Content-Type" content="text/html; charset=ISO-8859-1"/>
            <meta httpEquiv="X-UA-Compatible" content="IE=edge,chrome=1" />
            <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0"/>
            <title>{title ? title: ""}</title>
            {
            seoData.canonical_description && <meta name="description" content={seoData.canonical_description}/>
            }
            {
                seoData.canonical_seo_keywords && <meta name="keywords" content={seoData.canonical_seo_keywords} />
            }
            
            <meta name="robots" content="index,follow"/>
            {seoData.field_og_title && <meta property="og:title" content={seoData.field_og_title}/>}
            {
                seoData.field_og_description && <meta property="og:description" content={seoData.field_og_description}/>
            }
            {seoData.field_og_image_url && <meta property="og:image" content={seoData.field_og_image_url}/>} 
            {
                seoData.field_og_url && <meta property="og:url" content={seoData.field_og_url}/> 
            }
            <meta property="og:site_name" content="my webasite"/>
            <meta property="og:locale" content="en_US" />
            <meta property="og:type" content="website" />
            <meta name="twitter:card" content="summary"/>
            {
                seoData.field_twitter_title && <meta name="twitter:title" content={seoData.field_twitter_title}/>
            }
            {
                seoData.field_twitter_description && <meta name="twitter:description" content={seoData.field_twitter_description}/>
            }
            <meta name="twitter:domain" content="my website"/>
            {
                seoData.field_twitter_image_url && <meta name="twitter:image" content={seoData.field_twitter_image_url}/>
            }
            <meta property="twitter:site" content="@mysitezzz" />
            {
              seoData.field_itemprop_name && <meta itemProp="name" content={seoData.field_itemprop_name} />
            }
            {
             seoData.field_itemprop_description && <meta itemProp="description" content={seoData.field_itemprop_description} />
            }
            {
               seoData.field_itemprop_image_url && <meta itemProp="image" content={seoData.field_itemprop_image_url}/>
            }
            <meta property="fb:app_id" content="XXX"/>
            <meta name="author" content="my website"/>
            <meta name="apple-itunes-app" content="app-id=XXX"/>
            <meta name="google-play-app" content="app-id= XXX"/>
            <meta name="msApplication-ID" content="App"/>
            <meta name="msApplication-PackageFamilyName" content="XXX"/>
            <meta name="msapplication-TileImage" content="XXX"/>
            <meta name="HandheldFriendly" content="true"/>
            <meta httpEquiv="content-language" content="en-in"/>
            <meta name="geo.position" content="" />
            <meta name="geo.placename" content="" />
            <meta name="geo.region" content="" />
            <link rel="alternate" hrefLang="en" href="https://xyz.com/"/>
            {
                seoData.canonical_url && <link rel="canonical" href={seoData.canonical_url}/>
            }
            <link rel="alternate" href="android-app:// XXX" />
        </Head>
            <Header/>
            {children}
            <Footer/>
        </>
    ) 
}



export default BaseLayout;
```


I am able to see the meta tags while doing inspect element from the browser, but unable to see in the view source. 

Plus, I have a share functionality which needs to share the title and description from og:title, and og:description - Generally using the Open Graph tags to share it on Facebook, Twitter and Linkedin.

Facebook Debugger is scrapping the wrong data it seems. It is not setting that data which was populated to og:title and og:description meta tags


I am generally rendering the data on the client-side and that might be causing the issue as per my research, but what is the better solution on this??

Further Analysis -

I have the SEO reports from Lighthouse - currently showing a score of 85. SEO seems to be working perfectly for googlebots, but will not work with other crawlers - Such as Facebook for such instance and Facebook crawler showing every meta tags blank

**Comment by enricoschaaf** [comment]
If it's viable to generate this page at build time, use getStaticPaths and getStaticProps, if not use getServerSideProps, it's discouraged to use getInitalProps. https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation

**Reply by desaijay315** [reply]
**_I have used both getStaticProps and getServerSideProps, and It seems I am still facing the same issue._**

Also, while using getStaticProps - we need to generate the page always at the build time. what if there is a change in the CMS? This will not trigger changes at the request time. I think that's the caveats as mentioned in the docs as well.

**Reply by desaijay315** [reply]
**//implenting getStaticProps --**
```
export async function getStaticProps() {

    const seoData = await seoAPI();
  
    return {
        
        seoData
    }
}


export default Index;
```

**getServerSideProps**

```
export async function getServerSideProps(context) {

const seoData = await seoAPI();
  
    return {
        
        seoData
    }
}


export default Index;
```

**Comment by enricoschaaf** [comment, marked as answer]
This should definitely work, here an example https://github.com/enricoschaaf/meta-tags, https://meta-tags-58qlw6h98.vercel.app. And to answer your question, depending where you host your application you can set up a build hook with your cms, so whenever you change something in your CMS it will trigger a new build. And if your getStaticPaths has a fallback it will also try to generate routes at run time of they havent been rendered at build time.

**Reply by desaijay315** [reply]
I have tried for my application where the API is dynamically called and then set into the state.

So the initial state inside the BaseLayout component is undefined. So that canonical_description is undefined, then it gets the state-defined - n I think view source always show what is initially loaded to the HTML pages.

Can we async call the Head API, so that only once the data is defined, then only show the meta tags or something like that?

**Reply by desaijay315** [reply]
Also, the usage of functional or stateless function, does that matter ?

**Reply by desaijay315** [reply]
Hi @enricoschaaf , I have tried using your application and called the API, but the head content is still blank 

![image](https://user-images.githubusercontent.com/9349985/85039789-b43f7800-b1a5-11ea-9264-bd6fcfc714c7.png)

**Reply by enricoschaaf** [reply]
Maybe your API doesnt return any data?

**Reply by desaijay315** [reply]
API returns the data and it's used in the production env.

What I am seeing is -  One first render my data is blank array and then it get fill with proper data after i set the state. But I think, that's how it should be, right ? 

Check below - 


```


import React, { Component, PureComponent } from 'react';
import Head from 'next/head';



import {getHomePageData} from '../action/drupal-api-helper';
 
class  Index extends React.Component {
  // console.log(homepagedata);
  constructor(props){
    super(props);
    const {homepagedata } = props;
    // console.log(homepagedata);
    this.state = {
      field_paragraph_for_seo_opengrap:[]
    }
  }


    componentDidMount(){
      console.log(this.props.homepagedata);
      if(this.props.homepagedata && this.props.homepagedata.content && this.props.homepagedata.content.field_component_type && this.props.homepagedata.content.field_component_type.length > 0){
        const homeData = this.props.homepagedata.content.field_component_type.map((hData, index) => {
          if(hData.field_expertise_component){
            hData.field_title_expertise = hData.field_title;
          }else if(hData.field_stories_content_reference){
            hData.field_title_success = hData.field_title;
          }else if(hData.field_forward_thinking_component){
            hData.field_title_thinking = hData.field_title;
          }
          return hData;
        })
  
        for(const key in homeData){
          for(const keyd in homeData[key]){
            // console.log(keyd);
             this.state[keyd] = homeData[key][keyd];
          }
       }
       this.setState(this.state);     
      }
    }


  render(){
    console.log(this.state.field_paragraph_for_seo_opengrap);
    let url = this.state.field_paragraph_for_seo_opengrap && this.state.field_paragraph_for_seo_opengrap.length > 0 && this.state.field_paragraph_for_seo_opengrap[0].field_og_url ? this.state.field_paragraph_for_seo_opengrap[0].field_og_url: "";
    console.log(url);
    return (
      
      <Head>
        <meta property="og:url" content={url}/> 
      </Head>
    )
  }
}



export async function getStaticProps(context) {

  const homepagedata  = await getHomePageData();
  console.log(homepagedata);


  return {
      props:{
        homepagedata
      }
    }
}


export default Index;

```

**Reply by desaijay315** [reply]
**//action**

```
export const getHomePageData = async () => {
   return await axios.get(`http://fetch-api-data`).then(response => response.data);
}
```

**Reply by desaijay315** [reply]
**Console output** 

**_Note. - I can see this meta tag while doing inspect Element_**  -


![image](https://user-images.githubusercontent.com/9349985/85041566-bdc9df80-b1a7-11ea-82dc-c264964fb6cd.png)

**Reply by desaijay315** [reply]
@enricoschaaf - Here is the GITHUB REPO - [](https://github.com/desaijay315/meta-tags-nextjs/tree/master/main/website_live/my-app)

**Reply by desaijay315** [reply]
https://github.com/desaijay315/meta-tags-nextjs/tree/master/main/website_live/my-app

**Reply by enricoschaaf** [reply]
Well now it's obvious why it doesn't work, look at my example, you don't use any lifescycle events or state, you just accepted the prop and use it. As I did in my example. You basically tell next to just set the Meta Tags at runtime.

**Reply by desaijay315** [reply]
Can you please just paste the solution or error here?

**Reply by enricoschaaf** [reply]
Don't know what you mean with error.

**Reply by desaijay315** [reply]
I think I got that -
![image](https://user-images.githubusercontent.com/9349985/85045989-6890cc80-b1ad-11ea-8d62-6f6ae1aec277.png)

**Reply by desaijay315** [reply]
@enricoschaaf , Does this also mean that we must not set any data inside the state? I mean if u can just confirm on this?

**Reply by enricoschaaf** [reply]
If I understand you correctly, then yes you only need to use state for when you want to do something on the front end like toggeling a model or so.

**Reply by subgero** [reply]
hi @enricoschaaf your project doesnt exist anymore.

**Reply by ALi-Hosseinzad** [reply]
@enricoschaaf this link doesn't work (https://github.com/enricoschaaf/meta-tags)

**Reply by UIME** [reply]
(https://github.com/enricoschaaf/meta-tags) Also for me to

**Reply by Lukortech** [reply]
The link is no longer displaying any "install" button near the URL, so I am upping this.

**Comment by desaijay315** [comment]
Can i inject the Dynamic tags inside the _document.js directly for each page ?? Have any example for that?

**Comment by desaijay315** [comment]
I have my _document.js where I tried adding the meta tags and that started working. 

**Here is my code-**


```
import React from 'react'
import Document, { Head, Main, NextScript } from 'next/document'

import {GA_TRACKING_ID} from './lib';

export default class extends Document {
  render() {
    return (
      <html lang="en">
        <Head>
         <meta name="description" content={GA_TRACKING_ID}/>
        <meta name="keywords" content={GA_TRACKING_ID}/>
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </html>
    )
  }
}
```

`**`This does show in view source - `**
`

**Reply by enricoschaaf** [reply]
yes if every page should have the same meta description and keywords you can do that

**Reply by desaijay315** [reply]
Sorry, Every page has different meta description and keywords

**Reply by TRooYY** [reply]
so could you show the ./lib logic ? how to render different meta in different page? thanks

**Reply by desaijay315** [reply]
@TRooYY  - We can render them via the props received - 

For Example, If you are using getInitialprops or getStaticProps ..etc. Get the props for each page and render it out to a common meta.js component

**Reply by Guneetgstar** [reply]
@desaijay315 Where to save the common `meta.js`? How to render the `meta.js` for every page with dynamic props? Can you please show how it works with an example?

**Reply by subgero** [reply]
Hi @desaijay315 , did you solve this? I'm  facing the same problem, I'm using getServerSideProps, fetch data from an restful api, but the metadata returns empty. It would be nice if you elaborate your answer or if you share part of your code.

**Reply by subgero** [reply]
Did you manage to resolve this?

**Reply by hamirulhafizal** [reply]
Hello. any update?

**Reply by bbaymistery** [reply]
Guys did someone solve ?

**Reply by bbaymistery** [reply]
> - We can render them via the props received

Sir ?

**Reply by desaijay315** [reply]
Yes, What I was trying is to set the state and pass them to the meta.js component - This meta.js component can be a common component and then this received the actual props - Do not try to store the response into some state and then render

1. call getInitialProps /getStaticProps
2. response received is a props for the component in which  above two methods are called
3. pass this props to the meta component

Done!

**Reply by desaijay315** [reply]
> render

meta.js is nothing but the component which renders the meta tags (This can be dynamic). So each page will have different one

**Comment by artensonart98** [comment]
I don't know what's going under the hood, but removing the <Head> tag resolves this for me. I just put my meta tags without nesting them in Head, and it worked.

**Comment by bbaymistery** [comment]


**Reply by desaijay315** [reply]
At this point in time, i don't have the codebase, but if you can share the repo or GitHub with some dummy codebase will also do, i can then check and revert

**Comment by bbaymistery** [comment]


**Reply by desaijay315** [reply]
I can't see any meta tags here - Can you please redirect me to the file where you are using getStaticprops or ssr - and passing the props where you have dynamic meta tags ??

**Comment by bbaymistery** [comment]
https://stackoverflow.com/questions/62064804/view-page-source-does-not-show-page-in-nextjs/74078879#74078879   check this please .I managed to handle those So here is answer https://stackoverflow.com/questions/62064804/view-page-source-does-not-show-page-in-nextjs/74078879#74078879

**Comment by desaijay315** [comment]
I think then this should be fine

**Comment by sandrinjoy** [comment]
I just now faced a similar issue on one of my project, where the data is loaded on client side ...
After debugging for few hours, the issue was happing in my _app.js file, where i was having useeffect to create a fade in, fade out  animation effect using framer motion

```import "../styles/globals.scss";

import "./../config/config.styles.scss";
import {
  ThemeProvider,
  theme,
  CSSReset,
  ChakraProvider,
  ToastProvider,
} from "@chakra-ui/react";
import { chakraTheme } from "../config/chakra";
// import { ggSentry } from "../config/sentry";
// import { useRouter } from 'next/router';
// framer motion fade in animation for page transitions

import { motion } from "framer-motion";
import { useRouter } from "next/router";
import React, { useEffect, useState } from "react";
const SENTRY_KEY = "123";
function MyApp({ Component, pageProps }) {
  // ggSentry.initialize(SENTRY_KEY || process.env.REACT_APP_SENTRY);
  const getLayout = Component.getLayout || ((page) => page);

  // framer motion fade in animation for page transitions in next _app.js

  const router = useRouter();
  const [isMounted, setIsMounted] = useState(false);
  useEffect(() => {
    setIsMounted(true);
  }, []);

  return (
    <ThemeProvider theme={chakraTheme}>
      <ToastProvider>
        <CSSReset />
        {isMounted && (
          <motion.div
            key={router.route}
            initial="pageInitial"
            animate="pageAnimate"
            variants={{
              pageInitial: {
                opacity: 0,
              },
              pageAnimate: {
                opacity: 1,
              },
            }}
          >
            {getLayout(<Component {...pageProps} />)}
          </motion.div>
        )}
      </ToastProvider>
    </ThemeProvider>
  );
}

export default MyApp;
```

so the main issue here is the useeffect which will function only on the client side.. as well the my && condition check on for the transition .

The fix i could think of is that instead of && , toggling the opacity from 0 1 and vice-versa could be the right fix..


anyways the shortanswer is ,check if you are conditionally rendering your content which trigger after a useeffect

**Comment by khashayarghajar** [comment]
i'm facing the same issue

`<Head>
    <title>{Data?.Title}</title>
    {Data?.Meta?.map((tag, index) => (<meta key={index} {...tag} />))}
    {Data?.Link?.map((tag, index) => (<link key={index} {...tag} />))}
</Head>`

`export const getStaticProps: GetStaticProps = async () => {
  const Response: Response = await fetch(`${Server}/api/v1`);
  const Request: Request = await Response.json();
  return { props: { Data: Request }, revalidate: 120 };
};`

**Reply by bbaymistery** [reply]
https://stackoverflow.com/questions/62064804/view-page-source-does-not-show-page-in-nextjs/74078879#74078879   Check it out U should customize your document.js

**Reply by khashayarghajar** [reply]
thanks
